File: rust/src/parser/shape.rs
/// Shape detection: is a brace interior a σ-grammar alphabet, or a
/// concatenation of constructs (grouping brace)?
/// Mirrors himark/parser/_shape.py.

use crate::parser::phase2::{tokenize, RawNode};
use crate::parser::text::{split_top, strip_unescaped};

pub fn is_sequence_brace(content: &str) -> bool {
    // A bound `{floor:alphabet:ceiling}` is never a sequence brace.
    if split_top(":", content).len() == 3 {
        return false;
    }
    let body = if content.starts_with('!') {
        &content[1..]
    } else {
        content
    };
    for arm in split_top(",", body) {
        for part in split_top("..", &arm) {
            if !is_sigma_atom(&part) {
                return true;
            }
        }
    }
    false
}

fn is_sigma_atom(part: &str) -> bool {
    let part = strip_unescaped(part);
    let part = if part.starts_with('!') {
        part[1..].trim().to_string()
    } else {
        part
    };
    if part.is_empty() {
        return true;
    }
    let nodes = match tokenize(&part) {
        Ok(n) => n,
        Err(_) => return false,
    };
    let braces: Vec<_> = nodes
        .iter()
        .filter(|n| matches!(n, RawNode::Brace(_)))
        .collect();
    if braces.is_empty() {
        return true; // bare token
    }
    if braces.len() > 1 {
        return false;
    }
    // One brace — check it's not glued to adjacent text, and has at most exact count.
    let has_adjacent_text = nodes.iter().any(|n| match n {
        RawNode::Leaf(s) => !s.trim().is_empty(),
        _ => false,
    });
    if has_adjacent_text {
        return false;
    }
    if let Some(RawNode::Brace(b)) = braces.first() {
        if let Some(ref cs) = b.count_src {
            let cs = cs.trim();
            if !cs.chars().all(|c| c.is_ascii_digit()) {
                return false; // ranged/star count
            }
        }
    }
    true
}

File: himark/parser/_shape.py
"""Shape detection for a `{…}` brace interior.

Answers one question for phase3: is a brace's interior a single **σ-grammar
alphabet** expression (built from `..` and `,`), or a **concatenation** of
constructs — a grouping brace / sub-pattern (`{of{black}{quartz}}`)? This is a
pure predicate over the interior text; resolving the brace stays in phase3.
"""

import re

from himark.models import nodes_typed as t
from himark.parser import phase2
from himark.parser._text import brace_end, split_top, strip_unescaped


def is_sequence_brace(content: str) -> bool:
    """True if a brace's interior is a *grouping* — either a concatenation of
    constructs (`{of{black}{quartz}}`) or a single nested brace (`{{X}}`). Only a
    pure σ-expression (alphabet: `..`, `,`, `::` only) returns False.

    The σ-grammar has no concatenation operator: every `,`/`..`-separated part is
    bare text or exactly one `{…}` atom. A part that glues a construct onto
    adjacent text — or holds more than one construct — is a sub-pattern.
    A content that is itself one brace is a single-child grouping.
    """
    stripped = strip_unescaped(content)
    if stripped.startswith("{") and brace_end(stripped) == len(stripped):
        return True  # {{X}} — single-child grouping brace
    if len(split_top("::", content)) >= 2:
        return False  # a `{alphabet::floor..ceiling}` band is one value universe
    body = content
    if body.startswith("!"):
        body = body[1:]
    for arm in split_top(",", body):
        for part in split_top("..", arm):
            if not _is_sigma_atom(part):
                return True
    return False


def _is_sigma_atom(part: str) -> bool:
    """True if `part` is a valid σ atom: bare text, or a single `{…}` (optionally
    with an exact `[N]` count) surrounded only by whitespace. A construct glued
    to text, several constructs, or a ranged count is a sub-pattern fragment."""
    part = strip_unescaped(part)
    if part.startswith("!"):
        part = part[1:].strip()  # a `!` complement/exclusion arm, e.g. !{0,l,I,O}
    if not part:
        return True
    children = phase2.parse(part).children
    constructs = [c for c in children if isinstance(c, t.BraceGroupNode)]
    if not constructs:
        return True  # bare token (a..z, cat, etc.)
    if len(constructs) > 1:
        return False
    only = constructs[0]
    if any(isinstance(c, t.LeafNode) and c.content.strip() for c in children):
        return False  # a brace glued to adjacent literal text → concatenation
    if only.count_src is not None and not re.fullmatch(r"\d+", only.count_src.strip()):
        return False  # a ranged/star count is repetition, not a σ singleton
    return True
